lib/src/data/helper_classes/http_client.dart

```diff
@@ -0,0 +1,67 @@
+import 'dart:convert';
+import 'package:basic_authentication_flutter_challenge/injection.dart';
+import 'package:basic_authentication_flutter_challenge/src/data/constants/api_constants.dart';
+import 'package:basic_authentication_flutter_challenge/src/data/exceptions/http_exception.dart';
+import 'package:basic_authentication_flutter_challenge/src/domain/clients/api_client.dart';
+import 'package:basic_authentication_flutter_challenge/src/services/tokens_service.dart';
+import 'package:http/http.dart' as http;
+
+/// A helper class for wrapping the 3rd-party http package
+
+class HttpClient implements APIClient {
+  const HttpClient();
+
+  @override
+  Future<dynamic> get({
+    required String endPoint,
+    Map<String, dynamic>? query,
+  }) async {
+    final auth = locator<TokensService>();
+    Uri url = Uri.parse('$kApiBaseUrl$endPoint').replace(
+      queryParameters: query,
+    );
+    final headers = {
+      'Content-type': 'application/json',
+      'Accept': 'application/json',
+      'Authorization': await auth.getLocalAccessToken(),
+    };
+    final response = await http.get(url, headers: headers);
+    final body = jsonDecode(response.body);
+    final result = body['result'];
+    if (response.statusCode == 200) {
+      return result;
+    } else {
+      throw HttpException(
+        statusCode: response.statusCode,
+        message: result as String,
+      );
+    }
+  }
+
+  @override
+  Future<dynamic> post({
+    required String endPoint,
+    Map<String, dynamic>? query,
+  }) async {
+    final auth = locator<TokensService>();
+    Uri url = Uri.parse('$kApiBaseUrl$endPoint').replace(
+      queryParameters: query,
+    );
+    final headers = {
+      'Content-type': 'application/json',
+      'Accept': 'application/json',
+      'Authorization': await auth.getLocalAccessToken(),
+    };
+    final response = await http.post(url, headers: headers);
+    final body = jsonDecode(response.body);
+    final result = body['result'];
+    if (response.statusCode == 200) {
+      return result;
+    } else {
+      throw HttpException(
+        statusCode: response.statusCode,
+        message: result as String,
+      );
+    }
+  }
+}
```

Comment:
In both `get` & `post` functions your response handling should cover more edge cases. 

Your success validation has only one condition if the statusCode is 200. As you know we have different statusCode for success and for failures for example 201 for successfully created and 202 for acceptance. For more details check [mozilla website](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status).

I've experimented one-case to prove the concept. I changed the statusCode to 201 from client side to simulate statusCode incompatibility since I can't change it from server side.

As you can see it throws an exception even though I received a valid token.

![Screen Shot 1444-02-07 at 1 18 45 PM](https://user-images.githubusercontent.com/54819958/188266610-13b438c6-bc6d-4ffa-8d17-73376e9d38a0.png)

And the notifier notify the user with a technical error which need to be handled as well.

![Screen Shot 1444-02-07 at 1 20 42 PM](https://user-images.githubusercontent.com/54819958/188266686-aedacbba-0e0b-480e-8165-2e9de6d191a4.png)

Reply:
This should be done in coordination between the back end & front end developers. 
You should list all the status code that you use in your backend.

Reply:
For the notification, you are write, I'm just throwing everything at the user 😅

Reply:
I put my comments to tell you about specific things related to technical and soft skills. The good thing is you directly come to the point.

The takeaway here is about Backend and Frontend communication. You should enforce/convince backend developers to change backend code to minimize frontend logic and avoid hardcoding values. Believe you have to `push` them so hard. 

I know you already did #3 but didn't push me to change. It might be only a challenge and no harm to release with this code but never do it in real life.